File: apps/ai-service/app/api/ingest.py
from __future__ import annotations

from functools import lru_cache

from fastapi import APIRouter, Depends, File, UploadFile
from pydantic import BaseModel
from sqlalchemy import create_engine

from app.ai.embeddings import build_embedder
from app.ai.ingest import ingest_document
from app.ai.vectorstore import VectorStore
from app.api.triage import require_auth
from app.core.config import get_settings


@lru_cache
def _store() -> VectorStore:
    """Process-singleton store. Constructing it opens a SQLAlchemy engine (which
    owns a connection pool), so it must not be rebuilt per request. Uses the same
    embedder as the triage retriever so ingested vectors are searchable."""
    settings = get_settings()
    engine = create_engine(settings.database_url, pool_pre_ping=True)
    return VectorStore(engine, build_embedder(settings))


class IngestedFile(BaseModel):
    source: str
    chunks: int
    skipped: bool = False
    reason: str | None = None


class IngestResponse(BaseModel):
    files: list[IngestedFile]
    total_chunks: int


router = APIRouter(prefix="/ingest", tags=["ingest"], dependencies=[Depends(require_auth)])


@router.post("", response_model=IngestResponse)
async def ingest(files: list[UploadFile] = File(...)) -> IngestResponse:
    """Ingest one or more uploaded documents into the vector store.

    Multipart upload; each file is chunked and embedded. Binary and empty files
    are reported as skipped rather than failing the whole request. Auth is the
    same shared-secret JWT as the triage endpoint, so only callers holding a
    ticket-service token can add to the knowledge base."""
    store = _store()
    results: list[IngestedFile] = []
    total = 0
    for upload in files:
        raw = await upload.read()
        source = upload.filename or "upload"
        added, reason = ingest_document(store, source, raw)
        results.append(
            IngestedFile(source=source, chunks=added, skipped=added == 0, reason=reason)
        )
        total += added
    return IngestResponse(files=results, total_chunks=total)

File: apps/ai-service/internal/httpapi/ingest.go
package httpapi

import (
	"errors"
	"io"
	"mime/multipart"
	"net/http"

	"github.com/nickhildpac/ticket-management-ai-service/internal/rag"
)

// maxUploadBytes bounds the multipart body. It matches the cap the
// ticket-service ingest proxy already applies, so this is a defence in depth
// for direct callers.
const maxUploadBytes = 25 << 20 // 25 MiB

// ingestedFile is one file's outcome. reason is null on success, or
// "binary"/"empty" when nothing was stored.
type ingestedFile struct {
	Source  string  `json:"source"`
	Chunks  int     `json:"chunks"`
	Skipped bool    `json:"skipped"`
	Reason  *string `json:"reason"`
}

// ingestResponse is the payload the web admin UI renders (see
// apps/web/src/features/admin/queries.ts).
type ingestResponse struct {
	Files       []ingestedFile `json:"files"`
	TotalChunks int            `json:"total_chunks"`
}

// ingestHandler ingests one or more uploaded documents into the vector store.
//
// Multipart upload under the field name "files"; each file is chunked and
// embedded. Binary and empty files are reported as skipped rather than failing
// the whole request. Auth is the same shared-secret JWT as the triage endpoint,
// so only callers holding a ticket-service token can add to the knowledge base.
func ingestHandler(store *rag.VectorStore) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		r.Body = http.MaxBytesReader(w, r.Body, maxUploadBytes)
		reader, err := r.MultipartReader()
		if err != nil {
			writeError(w, http.StatusBadRequest, "validation_error",
				"expected a multipart/form-data upload", err.Error())
			return
		}

		resp := ingestResponse{Files: []ingestedFile{}}
		for {
			part, err := reader.NextPart()
			if errors.Is(err, io.EOF) {
				break
			}
			if err != nil {
				writeError(w, http.StatusBadRequest, "validation_error",
					"failed to read upload", err.Error())
				return
			}
			if part.FormName() != "files" {
				_ = part.Close()
				continue
			}

			file, err := ingestPart(r, store, part)
			_ = part.Close()
			if err != nil {
				writeError(w, http.StatusInternalServerError, "internal_server_error",
					"failed to ingest document", nil)
				return
			}
			resp.Files = append(resp.Files, file)
			resp.TotalChunks += file.Chunks
		}

		if len(resp.Files) == 0 {
			writeError(w, http.StatusBadRequest, "validation_error",
				"request validation failed",
				[]map[string]string{{
					"field": "files", "message": "field required", "type": "missing",
				}})
			return
		}
		writeJSON(w, http.StatusOK, resp)
	}
}

// ingestPart chunks and embeds a single uploaded file.
func ingestPart(r *http.Request, store *rag.VectorStore, part *multipart.Part) (ingestedFile, error) {
	source := part.FileName()
	if source == "" {
		source = "upload"
	}
	raw, err := io.ReadAll(part)
	if err != nil {
		return ingestedFile{}, err
	}

	added, reason, err := rag.IngestDocument(r.Context(), store, source, raw)
	if err != nil {
		return ingestedFile{}, err
	}

	out := ingestedFile{Source: source, Chunks: added, Skipped: added == 0}
	if reason != rag.SkipNone {
		text := string(reason)
		out.Reason = &text
	}
	return out, nil
}
